examples/login/src/main.rs

```diff
@@ -1,64 +1,243 @@
+            Ok(choice) => {
+                if choice >= choices.len() {
+                    eprintln!("This is not a valid choice");
+                } else {
+                    break choice;
+                }
             }
+            Err(_) => eprintln!("This is not a valid choice. Try again.\n"),
         };
+    };
+
+    choices[choice].login(client).await?;
+
+    Ok(())
+}
+
+/// Login with a username and password.
+async fn login_with_password(client: &Client) -> anyhow::Result<()> {
+    println!("Logging in with username and password…");
+
+    loop {
+        print!("\nUsername: ");
+        io::stdout().flush().expect("Unable to write to stdout");
+        let mut username = String::new();
+        io::stdin().read_line(&mut username).expect("Unable to read user input");
+        username = username.trim().to_owned();
+
+        print!("Password: ");
+        io::stdout().flush().expect("Unable to write to stdout");
+        let mut password = String::new();
+        io::stdin().read_line(&mut password).expect("Unable to read user input");
+        password = password.trim().to_owned();
+
+        match client
+            .login_username(&username, &password)
+            .initial_device_display_name(INITIAL_DEVICE_DISPLAY_NAME)
+            .await
+        {
+            Ok(_) => {
+                println!("Logged in as {username}");
+                break;
+            }
+            Err(error) => {
+                println!("Error logging in: {error}");
+                println!("Please try again\n");
+            }
+        }
+    }
 
-    login(homeserver_url, &username, &password).await?;
+    Ok(())
+}
+
+/// Login with SSO.
+async fn login_with_sso(client: &Client, idp: Option<&IdentityProvider>) -> anyhow::Result<()> {
+    println!("Logging in with SSO…");
+
+    let mut login_builder = client.login_sso(|url| async move {
+        // Usually we would want to use a library to open the URL in the browser, but
+        // let's keep it simple.
+        println!("\nOpen this URL in your browser: {url}\n");
+        println!("Waiting for login token…");
+        Ok(())
+    });
+
+    if let Some(idp) = idp {
+        login_builder = login_builder.identity_provider_id(&idp.id);
+    }
+
+    login_builder.await?;
+
+    println!("Logged in as {}", client.user_id().unwrap());
 
     Ok(())
 }
+
+/// Handle room messages by logging them.
+async fn on_room_message(event: OriginalSyncRoomMessageEvent, room: Room) {
+    // We only want to listen to joined rooms.
+    let Room::Joined(room) = room else {
+        return;
+    };
+
+    // We only want to log text messages.
+    let OriginalSyncRoomMessageEvent {
+            content:
+                RoomMessageEventContent {
+                    msgtype: MessageType::Text(TextMessageEventContent { body: msg_body, .. }),
+                    ..
+                },
+            sender,
+            ..
+        } = event else {
+            return;
+        };
```

Comment:
Indentation looks wrong here. Also I think it will be easier to read without struct literals, i.e. something like

```rust
let MessageType::Text(msgtype) = &event.content.msgtype else { return };
// use event.sender and msgtype.body
```